### ISSUE TITLE
Rule for JUnit tests should include assume.

### ISSUE THREAD

**Issue by Martin-Spamer** [pmd/pmd#684]
Please, prefix the report title with the language it applies to within brackets, such as *[java]* or *[apex]*. If not specific to a language, you can use *[core]*

**Rule Set:**
JUnit tests should include assert() or fail()

**Description:**
The Assume class provides valid pre-conditions for unit tests, to rule should not trigger if assume is present.  This should be treated in the same way as assert(...) and fail(...)

**Code Sample demonstrating the issue:**

```
    @Test 
    JUnit tests should include assert() or fail()
        public void testIsLocal() { 
            assumeTrue(TestContext.isLocal()); 
        } 
```

**Running PMD through:**
codacy

**Comment by jsotuyod** [comment]
@Martin-Spamer I haven't heard back from you. I'd love to get some insight into your use case.

Thanks in advanced

**Comment by Martin-Spamer** [comment]
Sorry for the delay,  The example was a minimal example to trigger the rule and I left the error in middle of the code.   I've clarified the original post to hopefully be more clear.  The specific justification for this change is as follows:

Yes, Assume.assumeTrue(...) is how JUnit does pre-conditions, it fulfils the role of {P} in a Hoare Triple, or a Given in BDD.  Its intended behaviour is to pass, or throw an assumption violation, not a normal fail. It will appear in test reports as a skipped test, not a failed test.  An assumption violation means the test will produce no meaningful result.  In this case 

http://junit.org/junit4/javadoc/4.12/org/junit/AssumptionViolatedException.html

I could dress up a unit test, make it not true and add a following fail(..) and followed by a fail, but that would require code changes to pass the rule only.  I don't want to disable this rule, it is a good one, I just expect it to pass with assume. 

I could do the following to pass PMD, but it obscures things and fails the principle of least surprise, and results in less clear code.
```
public class MyTestCase {
  @Test
  public void testIsLocal() { 
    assumeTrue(TestContext.isLocal()); 
  } 
  @Test
  public void testIsLocal() { 
    assumeFalse(TestContext.isLocal()); 
    fail("Not running locally...");
  } 
}
```

**Comment by jsotuyod** [comment]
@Martin-Spamer thanks for your time.

I fear I still don't see the point on having a test that can not fail, as those you provide in your example. As you said, assume works as the `given` in BDD, not as the `expect`. A test that can't fail is useless.

I believe there is an intrinsic problem on your tests, regardless of assumptions / PMD.

```java
  public void testIsLocal() { 
    assumeTrue(TestContext.isLocal()); 
  } 
```

Can only pass or be ignored. Will only fail when an exception is thrown in application code, and that holds true even without the assume:

```java
  public void testIsLocal() { 
    TestContext.isLocal(); 
  } 
```

This test can only pass, or fail upon an exception. I still see no point in having that assume there on your test case. Moreover, this second variant would also be flagged as missing an assertion, and that would still be a correct report. The correct implementation would be:

```java
  public void testIsLocal() { 
    try {
      TestContext.isLocal(); // I don't care if this returns true or false
    } catch (Exception e) {
      fail("no exceptions where expected");
    }
  } 
```

Preconditions ARE useful, but they are not test conditions. A test *can* be skipped if no meaningful result is to be expected, but **the precondition itself is not the test**. If the precondition holds, the test should still make proper assertions on the outcome of that test. For instance:

```java
public void testNewImplementation() {
  assumeTrue(isRunningJava8()); // check preconditions

  // Using Java 8, try our new implementation using lambdas
  assertTrue(myObj.useNewImplementation()); // validate our implementation is correct and holds
}
```

**Comment by pedrorijo91** [comment]
Hey, I have a similar case where we use assertj for assertions. I have a test method that relies on `assertThat` instead of `assertTrue/False`.

What's the best way to have PMD consider a wider range of assertion methods instead of just those 2?

**Comment by jsotuyod** [comment]
@pedrorijo91 supporting AssertJ seems like a different issue. Please open a new issue with a snippet of your failing code...

At first sight, using `assertThat(foo).isEqualTo(bar)` should work with the rule. However, `Assertions.assertThat(foo).isEqualTo(bar)` will probably fail...
